### ISSUE TITLE
Implement a generic class child of CharacterBody3D that implemens these stairs stepping behaviours

### ISSUE THREAD

**Issue by mrjshzk** [JheKWall/Godot-Stair-Step-Demo#2]
Hello and sorry to bother, this has been really helpful. Thank you :)
Currently only my player can stair-step, and I want my NPC's to also be able to do this the same way. Could this be possibly implemented?

**Comment by mrjshzk** [comment]
I will also try to do this myself, and I'll let you know my progress

**Comment by mrjshzk** [comment]
Okay, I just did it. :)

`extends CharacterBody3D
## Credits:
# Special thanks to Majikayo Games for original solution to stair_step_down!
# (https://youtu.be/-WjM1uksPIk)
#
# Special thanks to Myria666 for their paper on Quake movement mechanics (used for stair_step_up)!
# (https://github.com/myria666/qMovementDoc)
#
# Special thanks to Andicraft for their help with implementation stair_step_up!
# (https://github.com/Andicraft)
@onready var fv : Marker3D= $ForwardVector
@onready var visible_notifier : VisibleOnScreenNotifier3D= $VisibleOnScreenNotifier3D

@onready var navigation_agent : NavigationAgent3D= %NavigationAgent3D
@onready var player : CharacterBody3D = get_tree().get_first_node_in_group("Player")
@onready var navmesh : NavigationRegion3D = get_tree().get_first_node_in_group("Navmesh")
## Notes:
# 0. All shape colliders are supported. Although, I would recommend Capsule colliders for enemies
#		as it works better with the Navigation Meshes. Its up to you what shape you want to use
#		for players.
#
# 1. To adjust the step-up/down height, just change the MAX_STEP_UP/MAX_STEP_DOWN values below.
#
# 2. This uses Jolt Physics as the default Godot Physics has a few bugs:
#	1: Small gaps that you should be able to fit through both ways will block you in Godot Physics.
#		You can see this demonstrated with the floating boxes in front of the big stairs.
#	2: Walking into some objects may push the player downward by a small amount which causes
#		jittering and causes the floor to be detected as a step.
#	TLDR: This still works with default Godot Physics, although it feels a lot better in Jolt Physics.

#region ANNOTATIONS ################################################################################
@export_category("Player Settings")
@export var PLAYER_SPEED := 10.0		# Player's movement speed.
@export var JUMP_VELOCITY := 6.0		# Player's jump velocity.

@export var MAX_STEP_UP := 0.5			# Maximum height in meters the player can step up.
@export var MAX_STEP_DOWN := -0.5		# Maximum height in meters the player can step down.


@export_category("Debug Settings")
@export var STEP_DOWN_DEBUG := false	# Enable these to get detailed info on the step down/up process.
@export var STEP_UP_DEBUG := false

## Node References
@onready var PLAYER_COLLIDER = $CollisionShape3D


#endregion

#region VARIABLES ##################################################################################
var is_grounded := true					# If player is grounded this frame
var was_grounded := true				# If player was grounded last frame

var vertical := Vector3(0, 1, 0)		# Shortcut for converting vectors to vertical
var horizontal := Vector3(1, 0, 1)		# Shortcut for converting vectors to horizontal

var gravity = ProjectSettings.get_setting("physics/3d/default_gravity")		# Default Gravity
#endregion

func _ready() -> void:
	navigation_agent.velocity_computed.connect(_on_velocity_computed)
	set_movement_target(Vector3(randf_range(-30,30), randf_range(-30,30), randf_range(-30,30)) + global_position)

func set_movement_target(movement_target: Vector3):
	navigation_agent.set_target_position(movement_target)

func smooth_look_at(pos : Vector3, smoothness: float):
	var fw := pos - global_position
	if fw.cross(Vector3(0,1,0)).is_zero_approx() or fw == Vector3.ZERO: return
	var lookat_basis := Basis.looking_at(fw, Vector3(0,1,0), true)
	var original_scale := basis.get_scale()
	
	basis = basis.orthonormalized()
	var slerped_basis = basis.slerp(lookat_basis, get_physics_process_delta_time() * smoothness)
	global_transform = Transform3D(slerped_basis, global_position)
	scale = original_scale
	rotation_degrees.x = 0
	rotation_degrees.z = 0

func _on_velocity_computed(safe_velocity: Vector3):
	print(safe_velocity)
	velocity = safe_velocity
	move_and_slide()
	# Stair step down
	stair_step_down()


# Function: Handle frame-based physics processes
func _physics_process(delta):
	# Lock player collider rotation
	PLAYER_COLLIDER.global_rotation = Vector3.ZERO

	# Update player state
	was_grounded = is_grounded

	if is_on_floor():
		is_grounded = true
	else:
		is_grounded = false

	# Handle Gravity
	if !is_on_floor():
		velocity.y -= gravity * delta

	smooth_look_at(navigation_agent.get_next_path_position(), 5)
	
	var next_path_position: Vector3 = navigation_agent.get_next_path_position()
	
	var new_velocity: Vector3 = global_position.direction_to(next_path_position)
	
	# Stair step up
	stair_step_up()

	# Move
	navigation_agent.set_velocity(new_velocity * PLAYER_SPEED)


# Function: Handle walking down stairs
func stair_step_down():
	if is_grounded:
		return

	# If we're falling from a step
	if velocity.y <= 0 and was_grounded:
		_debug_stair_step_down("SSD_ENTER", null)													## DEBUG

		# Initialize body test variables
		var body_test_result = PhysicsTestMotionResult3D.new()
		var body_test_params = PhysicsTestMotionParameters3D.new()

		body_test_params.from = self.global_transform			## We get the player's current global_transform
		body_test_params.motion = Vector3(0, MAX_STEP_DOWN, 0)	## We project the player downward

		if PhysicsServer3D.body_test_motion(self.get_rid(), body_test_params, body_test_result):
			# Enters if a collision is detected by body_test_motion
			# Get distance to step and move player downward by that much
			position.y += body_test_result.get_travel().y
			apply_floor_snap()
			is_grounded = true
			_debug_stair_step_down("SSD_APPLIED", body_test_result.get_travel().y)					## DEBUG

# Function: Handle walking up stairs
func stair_step_up():
	_debug_stair_step_up("SSU_ENTER", null)															## DEBUG

	# 0. Initialize testing variables
	var body_test_params = PhysicsTestMotionParameters3D.new()
	var body_test_result = PhysicsTestMotionResult3D.new()

	var test_transform = global_transform				## Storing current global_transform for testing
	var distance = Vector3(velocity.x, 0, velocity.z).normalized() * 0.1						## Distance forward we want to check
	body_test_params.from = self.global_transform		## Self as origin point
	body_test_params.motion = distance					## Go forward by current distance

	_debug_stair_step_up("SSU_TEST_POS", test_transform)											## DEBUG

	# Pre-check: Are we colliding?
	if !PhysicsServer3D.body_test_motion(self.get_rid(), body_test_params, body_test_result):
		_debug_stair_step_up("SSU_EXIT_1", null)													## DEBUG

		## If we don't collide, return
		return

	# 1. Move test_transform to collision location
	var remainder = body_test_result.get_remainder()							## Get remainder from collision
	test_transform = test_transform.translated(body_test_result.get_travel())	## Move test_transform by distance traveled before collision

	_debug_stair_step_up("SSU_REMAINING", remainder)												## DEBUG
	_debug_stair_step_up("SSU_TEST_POS", test_transform)											## DEBUG

	# 2. Move test_transform up to ceiling (if any)
	var step_up = MAX_STEP_UP * vertical
	body_test_params.from = test_transform
	body_test_params.motion = step_up
	PhysicsServer3D.body_test_motion(self.get_rid(), body_test_params, body_test_result)
	test_transform = test_transform.translated(body_test_result.get_travel())

	_debug_stair_step_up("SSU_TEST_POS", test_transform)											## DEBUG

	# 3. Move test_transform forward by remaining distance
	body_test_params.from = test_transform
	body_test_params.motion = remainder
	PhysicsServer3D.body_test_motion(self.get_rid(), body_test_params, body_test_result)
	test_transform = test_transform.translated(body_test_result.get_travel())

	_debug_stair_step_up("SSU_TEST_POS", test_transform)											## DEBUG

	# 3.5 Project remaining along wall normal (if any)
	## So you can walk into wall and up a step
	if body_test_result.get_collision_count() != 0:
		remainder = body_test_result.get_remainder().length()

		### Uh, there may be a better way to calculate this in Godot.
		var wall_normal = body_test_result.get_collision_normal()
		var dot_div_mag = Vector3(velocity.x, 0, velocity.z).normalized().dot(wall_normal) / (wall_normal * wall_normal).length()
		var projected_vector = (Vector3(velocity.x, 0, velocity.z).normalized() - dot_div_mag * wall_normal).normalized()

		body_test_params.from = test_transform
		body_test_params.motion = remainder * projected_vector
		PhysicsServer3D.body_test_motion(self.get_rid(), body_test_params, body_test_result)
		test_transform = test_transform.translated(body_test_result.get_travel())

		_debug_stair_step_up("SSU_TEST_POS", test_transform)										## DEBUG

	# 4. Move test_transform down onto step
	body_test_params.from = test_transform
	body_test_params.motion = MAX_STEP_UP * -vertical

	# Return if no collision
	if !PhysicsServer3D.body_test_motion(self.get_rid(), body_test_params, body_test_result):
		_debug_stair_step_up("SSU_EXIT_2", null)													## DEBUG

		return

	test_transform = test_transform.translated(body_test_result.get_travel())
	_debug_stair_step_up("SSU_TEST_POS", test_transform)											## DEBUG

	# 5. Check floor normal for un-walkable slope
	var surface_normal = body_test_result.get_collision_normal()
	print("SSU: Surface check: ", snappedf(surface_normal.angle_to(vertical), 0.001), " vs ", floor_max_angle)#!
	if (snappedf(surface_normal.angle_to(vertical), 0.001) > floor_max_angle):
		_debug_stair_step_up("SSU_EXIT_3", null)													## DEBUG

		return
	print("SSU: Walkable")#!
	_debug_stair_step_up("SSU_TEST_POS", test_transform)											## DEBUG

	# 6. Move player up
	var global_pos = global_position
	var step_up_dist = test_transform.origin.y - global_pos.y
	_debug_stair_step_up("SSU_APPLIED", step_up_dist)												## DEBUG

	velocity.y = 0
	global_pos.y = test_transform.origin.y
	global_position = global_pos

## Debugging #######################################################################################

# Debug: Stair Step Down
func _debug_stair_step_down(param, value):
	if STEP_DOWN_DEBUG == false:
		return

	match param:
		"SSD_ENTER":
			print()
			print("Stair step down entered")
		"SSD_APPLIED":
			print("Stair step down applied, travel = ", value)

# Debug: Stair Step Up
func _debug_stair_step_up(param, value):
	if STEP_UP_DEBUG == false:
		return

	match param:
		"SSU_ENTER":
			print()
			print("SSU: Stair step up entered")
		"SSU_EXIT_1":
			print("SSU: Exited with no collisions")
		"SSU_TEST_POS":
			print("SSU: test_transform current position = ", value)
		"SSU_REMAINING":
			print("SSU: Remaining distance = ", value)
		"SSU_EXIT_2":
			print("SSU: Exited due to no step collision")
		"SSU_EXIT_3":
			print("SSU: Exited due to non-floor stepping")
		"SSU_APPLIED":
			print("SSU: Player moved up by ", value, " units")



#endregion

#region SIGNALS ####################################################################################
# Button: Change MAX_STEP_UP/MAX_STEP_DOWN to 0.5/-0.5
func _on_step_0_5_pressed():
	MAX_STEP_UP = 0.5
	MAX_STEP_DOWN = -0.5

# Button: Change MAX_STEP_UP/MAX_STEP_DOWN to 1/-1
func _on_step_1_pressed():
	MAX_STEP_UP = 1
	MAX_STEP_DOWN = -1

# Button: Change MAX_STEP_UP/MAX_STEP_DOWN to 2/-2
func _on_step_2_pressed():
	MAX_STEP_UP = 2
	MAX_STEP_DOWN = -2

# Button: Change MAX_STEP_UP/MAX_STEP_DOWN to 4/-4
func _on_step_4_pressed():
	MAX_STEP_UP = 4
	MAX_STEP_DOWN = -4

# Button: Change MAX_STEP_UP/MAX_STEP_DOWN to 100/-100
func _on_step_100_pressed():
	MAX_STEP_UP = 100
	MAX_STEP_DOWN = -100

#endregion


func _on_navigation_agent_3d_navigation_finished():
	set_movement_target(Vector3(randf_range(-30,30), randf_range(-30,30), randf_range(-30,30)))
	print("Enemy navigation finished.")
`
